File: packages/ccode/src/observer/dial.ts
/**
 * Dial Control Types for Observer Network
 *
 * Provides independent control over three dimensions:
 * - Observe: 0% = passive wait, 100% = active scanning
 * - Decide: 0% = suggest only, 100% = autonomous decision
 * - Act: 0% = wait for confirmation, 100% = immediate execution
 *
 * @module observer/dial
 */

import { Log } from "@/util/log"

const log = Log.create({ service: "observer.dial" })

// ─────────────────────────────────────────────────────────────────────────────
// Types
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Dial operating mode.
 */
export type DialMode = "manual" | "adaptive"

/**
 * Gear presets for intuitive control.
 *
 * Like a car's gear selector:
 * - P (Park): Everything off, no resource consumption
 * - N (Neutral): Observe only, no intervention
 * - D (Drive): Balanced autonomy for daily use
 * - S (Sport): High autonomy, aggressive mode
 * - M (Manual): Full user control over each dial
 */
export type GearPreset = "P" | "N" | "D" | "S" | "M"

/**
 * Dial names for type safety.
 */
export type DialName = "observe" | "decide" | "act"

/**
 * Configuration for a single dial.
 */
export interface DialConfig {
  /** Current value (0-100) */
  value: number
  /** Operating mode */
  mode: DialMode
  /** Bounds for adaptive mode */
  bounds: {
    min: number
    max: number
  }
}

/**
 * Three dials configuration.
 */
export interface ThreeDialsConfig {
  observe: DialConfig
  decide: DialConfig
  act: DialConfig
}

/**
 * Simple dial values (without mode and bounds).
 */
export interface DialValues {
  observe: number
  decide: number
  act: number
}

// ─────────────────────────────────────────────────────────────────────────────
// Gear Presets
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Preset dial values for each gear.
 */
export const GEAR_PRESETS: Record<GearPreset, DialValues> = {
  P: { observe: 0, decide: 0, act: 0 },     // Park: everything off
  N: { observe: 50, decide: 0, act: 0 },    // Neutral: observe only
  D: { observe: 70, decide: 60, act: 40 },  // Drive: balanced autonomy
  S: { observe: 90, decide: 80, act: 70 },  // Sport: high autonomy
  M: { observe: 50, decide: 50, act: 50 },  // Manual: neutral starting point
}

/**
 * Gear metadata for display.
 */
export const GEAR_INFO: Record<GearPreset, { name: string; description: string }> = {
  P: { name: "Park", description: "System inactive, no resource consumption" },
  N: { name: "Neutral", description: "Observe and record only, no intervention" },
  D: { name: "Drive", description: "Balanced autonomy for daily operation" },
  S: { name: "Sport", description: "High autonomy, aggressive mode" },
  M: { name: "Manual", description: "Full manual control over each dial" },
}

// ─────────────────────────────────────────────────────────────────────────────
// Dial Class
// ─────────────────────────────────────────────────────────────────────────────

/**
 * A single dial with value, mode, and adaptive bounds.
 */
export class Dial {
  private _value: number
  private _mode: DialMode
  private _min: number
  private _max: number

  constructor(value = 50, mode: DialMode = "manual", min = 0, max = 100) {
    this._value = this.clamp(value, 0, 100)
    this._mode = mode
    this._min = this.clamp(min, 0, 100)
    this._max = this.clamp(max, this._min, 100)
  }

  /** Get current value. */
  get value(): number {
    return this._value
  }

  /** Set value, respecting bounds in adaptive mode. */
  set value(v: number) {
    const clamped = this.clamp(v, 0, 100)
    this._value = this._mode === "adaptive"
      ? this.clamp(clamped, this._min, this._max)
      : clamped
  }

  /** Get operating mode. */
  get mode(): DialMode {
    return this._mode
  }

  /** Set operating mode. */
  set mode(m: DialMode) {
    this._mode = m
    if (m === "adaptive") {
      this._value = this.clamp(this._value, this._min, this._max)
    }
  }

  /** Get bounds. */
  get bounds(): { min: number; max: number } {
    return { min: this._min, max: this._max }
  }

  /** Set bounds (for adaptive mode). */
  setBounds(min: number, max: number): void {
    this._min = this.clamp(min, 0, 100)
    this._max = this.clamp(max, this._min, 100)
    if (this._mode === "adaptive") {
      this._value = this.clamp(this._value, this._min, this._max)
    }
  }

  /** Adjust value by delta. */
  adjust(delta: number): void {
    this.value = this._value + delta
  }

  /** Check if dial is active (value > 0). */
  isActive(): boolean {
    return this._value > 0
  }

  /** Check if dial is above threshold (default 50). */
  isHigh(threshold = 50): boolean {
    return this._value > threshold
  }

  /** Get value as float (0.0 - 1.0). */
  asFloat(): number {
    return this._value / 100
  }

  /** Switch to manual mode. */
  toManual(): void {
    this._mode = "manual"
  }

  /** Switch to adaptive mode with bounds. */
  toAdaptive(min: number, max: number): void {
    this._mode = "adaptive"
    this.setBounds(min, max)
  }

  /** Get config object. */
  toConfig(): DialConfig {
    return {
      value: this._value,
      mode: this._mode,
      bounds: { min: this._min, max: this._max },
    }
  }

  /** Create from config. */
  static fromConfig(config: DialConfig): Dial {
    return new Dial(config.value, config.mode, config.bounds.min, config.bounds.max)
  }

  private clamp(value: number, min: number, max: number): number {
    return Math.max(min, Math.min(max, Math.round(value)))
  }
}

// ─────────────────────────────────────────────────────────────────────────────
// ThreeDials Class
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Three independent dials for observe/decide/act control.
 */
export class ThreeDials {
  /** Observation intensity: 0% = passive, 100% = active scanning */
  readonly observe: Dial
  /** Decision autonomy: 0% = suggest only, 100% = autonomous */
  readonly decide: Dial
  /** Execution autonomy: 0% = wait for confirmation, 100% = immediate */
  readonly act: Dial

  /** Current gear (null if custom values don't match any preset) */
  private _gear: GearPreset | null

  constructor(observe = 70, decide = 60, act = 40, gear: GearPreset | null = null) {
    this.observe = new Dial(observe)
    this.decide = new Dial(decide)
    this.act = new Dial(act)
    this._gear = gear ?? this.detectGear()
  }

  /** Get current gear. */
  get gear(): GearPreset | null {
    return this._gear
  }

  /** Create from a gear preset. */
  static fromGear(gear: GearPreset): ThreeDials {
    const values = GEAR_PRESETS[gear]
    return new ThreeDials(values.observe, values.decide, values.act, gear)
  }

  /** Create with custom values. */
  static custom(observe: number, decide: number, act: number): ThreeDials {
    return new ThreeDials(observe, decide, act, "M")
  }

  /** Create from config. */
  static fromConfig(config: ThreeDialsConfig): ThreeDials {
    const dials = new ThreeDials()
    Object.assign(dials.observe, Dial.fromConfig(config.observe))
    Object.assign(dials.decide, Dial.fromConfig(config.decide))
    Object.assign(dials.act, Dial.fromConfig(config.act))
    return dials
  }

  /** Set gear preset. */
  setGear(gear: GearPreset, customValues?: DialValues): void {
    if (gear === "M" && customValues) {
      this.observe.value = customValues.observe
      this.decide.value = customValues.decide
      this.act.value = customValues.act
    } else {
      const values = GEAR_PRESETS[gear]
      this.observe.value = values.observe
      this.decide.value = values.decide
      this.act.value = values.act
    }
    this._gear = gear

    log.debug("Gear set", { gear, values: this.values() })
  }

  /** Set a single dial and switch to Manual mode. */
  setDial(name: DialName, value: number): void {
    switch (name) {
      case "observe":
        this.observe.value = value
        break
      case "decide":
        this.decide.value = value
        break
      case "act":
        this.act.value = value
        break
    }
    this._gear = "M" // Switch to manual when individual dial is adjusted

    log.debug("Dial set", { dial: name, value, gear: this._gear })
  }

  /** Get dial by name. */
  getDial(name: DialName): Dial {
    switch (name) {
      case "observe":
        return this.observe
      case "decide":
        return this.decide
      case "act":
        return this.act
    }
  }

  /** Check if observation should be active. */
  shouldObserve(): boolean {
    return this.observe.isActive()
  }

  /** Check if observation is in high/active mode. */
  isObservingActively(): boolean {
    return this.observe.isHigh()
  }

  /** Check if decisions should be made autonomously. */
  shouldDecideAutonomously(): boolean {
    return this.decide.isHigh()
  }

  /** Check if actions should be executed immediately. */
  shouldActImmediately(): boolean {
    return this.act.isHigh()
  }

  /** Check if all dials are off (Park mode). */
  isParked(): boolean {
    return !this.observe.isActive() && !this.decide.isActive() && !this.act.isActive()
  }

  /** Get all values as object. */
  values(): DialValues {
    return {
      observe: this.observe.value,
      decide: this.decide.value,
      act: this.act.value,
    }
  }

  /** Get combined autonomy score (average of all dials). */
  autonomyScore(): number {
    return Math.round((this.observe.value + this.decide.value + this.act.value) / 3)
  }

  /** Get config object. */
  toConfig(): ThreeDialsConfig {
    return {
      observe: this.observe.toConfig(),
      decide: this.decide.toConfig(),
      act: this.act.toConfig(),
    }
  }

  /** Detect which gear preset matches current values (if any). */
  private detectGear(): GearPreset | null {
    const v = this.values()
    for (const [gear, preset] of Object.entries(GEAR_PRESETS) as [GearPreset, DialValues][]) {
      if (v.observe === preset.observe && v.decide === preset.decide && v.act === preset.act) {
        return gear
      }
    }
    return null
  }
}

// ─────────────────────────────────────────────────────────────────────────────
// Factory Functions
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Create a ThreeDials instance from a gear preset.
 */
export function createDialsFromGear(gear: GearPreset): ThreeDials {
  return ThreeDials.fromGear(gear)
}

/**
 * Create a ThreeDials instance with custom values.
 */
export function createCustomDials(observe: number, decide: number, act: number): ThreeDials {
  return ThreeDials.custom(observe, decide, act)
}

/**
 * Parse a gear string to GearPreset.
 */
export function parseGear(input: string): GearPreset | null {
  const normalized = input.toUpperCase().trim()
  switch (normalized) {
    case "P":
    case "PARK":
      return "P"
    case "N":
    case "NEUTRAL":
      return "N"
    case "D":
    case "DRIVE":
      return "D"
    case "S":
    case "SPORT":
      return "S"
    case "M":
    case "MANUAL":
      return "M"
    default:
      return null
  }
}

/**
 * Check if a gear allows autonomous decisions.
 */
export function gearAllowsAutonomousDecisions(gear: GearPreset): boolean {
  return gear !== "P" && gear !== "N"
}

/**
 * Check if a gear allows autonomous actions.
 */
export function gearAllowsAutonomousActions(gear: GearPreset): boolean {
  return gear === "S"
}

File: services/zero-hub/src/observer/dial.rs
//! Three-dial control system for the Observer Network.
//!
//! Provides independent control over three dimensions:
//! - **Observe**: 0% = passive wait, 100% = active scanning
//! - **Decide**: 0% = suggest only, 100% = autonomous decision
//! - **Act**: 0% = wait for confirmation, 100% = immediate execution
//!
//! Each dial can operate in two modes:
//! - **Manual**: Fixed value set by user
//! - **Adaptive**: Value adjusts within bounds based on context

use serde::{Deserialize, Serialize};
use std::fmt;

/// Dial operating mode.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum DialMode {
    /// Fixed value, does not change automatically
    Manual,
    /// Value adjusts within bounds based on context
    Adaptive,
}

impl Default for DialMode {
    fn default() -> Self {
        Self::Manual
    }
}

impl fmt::Display for DialMode {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Manual => write!(f, "manual"),
            Self::Adaptive => write!(f, "adaptive"),
        }
    }
}

/// A single dial with value, mode, and adaptive bounds.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Dial {
    /// Current value (0-100)
    pub value: u8,
    /// Operating mode (manual or adaptive)
    pub mode: DialMode,
    /// Minimum value for adaptive mode
    pub min: u8,
    /// Maximum value for adaptive mode
    pub max: u8,
}

impl Dial {
    /// Create a new dial with a fixed value.
    pub fn new(value: u8) -> Self {
        Self {
            value: value.min(100),
            mode: DialMode::Manual,
            min: 0,
            max: 100,
        }
    }

    /// Create an adaptive dial with bounds.
    pub fn adaptive(value: u8, min: u8, max: u8) -> Self {
        let min = min.min(100);
        let max = max.min(100).max(min);
        let value = value.min(100).clamp(min, max);

        Self {
            value,
            mode: DialMode::Adaptive,
            min,
            max,
        }
    }

    /// Set the dial value, respecting bounds in adaptive mode.
    pub fn set(&mut self, value: u8) {
        let value = value.min(100);
        self.value = match self.mode {
            DialMode::Manual => value,
            DialMode::Adaptive => value.clamp(self.min, self.max),
        };
    }

    /// Adjust the dial by a delta amount.
    pub fn adjust(&mut self, delta: i16) {
        let new_value = (self.value as i16 + delta).clamp(0, 100) as u8;
        self.set(new_value);
    }

    /// Check if the dial is active (value > 0).
    pub fn is_active(&self) -> bool {
        self.value > 0
    }

    /// Check if the dial is above threshold (default 50).
    pub fn is_high(&self) -> bool {
        self.value > 50
    }

    /// Check if the dial is at or above a custom threshold.
    pub fn above(&self, threshold: u8) -> bool {
        self.value >= threshold
    }

    /// Get value as a float (0.0 - 1.0).
    pub fn as_float(&self) -> f32 {
        self.value as f32 / 100.0
    }

    /// Switch to manual mode with current value.
    pub fn to_manual(&mut self) {
        self.mode = DialMode::Manual;
    }

    /// Switch to adaptive mode with bounds.
    pub fn to_adaptive(&mut self, min: u8, max: u8) {
        self.mode = DialMode::Adaptive;
        self.min = min.min(100);
        self.max = max.min(100).max(self.min);
        self.value = self.value.clamp(self.min, self.max);
    }
}

impl Default for Dial {
    fn default() -> Self {
        Self::new(50)
    }
}

impl fmt::Display for Dial {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self.mode {
            DialMode::Manual => write!(f, "{}%", self.value),
            DialMode::Adaptive => write!(f, "{}% [{}-{}]", self.value, self.min, self.max),
        }
    }
}

/// Three independent dials for observe/decide/act control.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ThreeDials {
    /// Observation intensity: 0% = passive, 100% = active scanning
    pub observe: Dial,
    /// Decision autonomy: 0% = suggest only, 100% = autonomous
    pub decide: Dial,
    /// Execution autonomy: 0% = wait for confirmation, 100% = immediate
    pub act: Dial,
}

impl ThreeDials {
    /// Create dials with specific values.
    pub fn new(observe: u8, decide: u8, act: u8) -> Self {
        Self {
            observe: Dial::new(observe),
            decide: Dial::new(decide),
            act: Dial::new(act),
        }
    }

    /// Create dials with custom values (alias for new).
    pub fn custom(observe: u8, decide: u8, act: u8) -> Self {
        Self::new(observe, decide, act)
    }

    /// Create dials from a gear preset.
    pub fn from_gear(gear: super::gear::Gear) -> Self {
        use super::gear::Gear;

        match gear {
            Gear::P => Self::new(0, 0, 0),     // Park: everything off
            Gear::N => Self::new(50, 0, 0),    // Neutral: observe only
            Gear::D => Self::new(70, 60, 40),  // Drive: balanced autonomy
            Gear::S => Self::new(90, 80, 70),  // Sport: high autonomy
            Gear::M => Self::new(50, 50, 50),  // Manual: neutral starting point
        }
    }

    /// Check if observation should be active.
    pub fn should_observe(&self) -> bool {
        self.observe.is_active()
    }

    /// Check if observation is in high/active mode.
    pub fn is_observing_actively(&self) -> bool {
        self.observe.is_high()
    }

    /// Check if decisions should be made autonomously.
    pub fn should_decide_autonomously(&self) -> bool {
        self.decide.is_high()
    }

    /// Check if actions should be executed immediately.
    pub fn should_act_immediately(&self) -> bool {
        self.act.is_high()
    }

    /// Check if all dials are off (Park mode).
    pub fn is_parked(&self) -> bool {
        !self.observe.is_active() && !self.decide.is_active() && !self.act.is_active()
    }

    /// Get all values as a tuple.
    pub fn values(&self) -> (u8, u8, u8) {
        (self.observe.value, self.decide.value, self.act.value)
    }

    /// Get combined autonomy score (average of all dials).
    pub fn autonomy_score(&self) -> u8 {
        ((self.observe.value as u16 + self.decide.value as u16 + self.act.value as u16) / 3) as u8
    }

    /// Check if current values match a gear preset.
    pub fn matches_gear(&self) -> Option<super::gear::Gear> {
        use super::gear::Gear;

        let (o, d, a) = self.values();

        if o == 0 && d == 0 && a == 0 {
            Some(Gear::P)
        } else if o == 50 && d == 0 && a == 0 {
            Some(Gear::N)
        } else if o == 70 && d == 60 && a == 40 {
            Some(Gear::D)
        } else if o == 90 && d == 80 && a == 70 {
            Some(Gear::S)
        } else {
            None // Manual or custom
        }
    }
}

impl Default for ThreeDials {
    fn default() -> Self {
        Self::from_gear(super::gear::Gear::D)
    }
}

impl fmt::Display for ThreeDials {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "Observe: {}, Decide: {}, Act: {}",
            self.observe, self.decide, self.act
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_dial_new() {
        let dial = Dial::new(75);
        assert_eq!(dial.value, 75);
        assert_eq!(dial.mode, DialMode::Manual);
    }

    #[test]
    fn test_dial_clamps_to_100() {
        let dial = Dial::new(150);
        assert_eq!(dial.value, 100);
    }

    #[test]
    fn test_dial_adaptive() {
        let dial = Dial::adaptive(50, 30, 80);
        assert_eq!(dial.value, 50);
        assert_eq!(dial.min, 30);
        assert_eq!(dial.max, 80);
        assert_eq!(dial.mode, DialMode::Adaptive);
    }

    #[test]
    fn test_dial_adaptive_clamps() {
        let mut dial = Dial::adaptive(50, 30, 80);

        dial.set(90); // Should clamp to max
        assert_eq!(dial.value, 80);

        dial.set(10); // Should clamp to min
        assert_eq!(dial.value, 30);
    }

    #[test]
    fn test_dial_adjust() {
        let mut dial = Dial::new(50);

        dial.adjust(20);
        assert_eq!(dial.value, 70);

        dial.adjust(-30);
        assert_eq!(dial.value, 40);

        dial.adjust(-100);
        assert_eq!(dial.value, 0);

        dial.adjust(200);
        assert_eq!(dial.value, 100);
    }

    #[test]
    fn test_dial_thresholds() {
        let low = Dial::new(30);
        let high = Dial::new(70);
        let zero = Dial::new(0);

        assert!(!low.is_high());
        assert!(high.is_high());
        assert!(!zero.is_active());
        assert!(low.is_active());
    }

    #[test]
    fn test_three_dials_from_gear() {
        use super::super::gear::Gear;

        let park = ThreeDials::from_gear(Gear::P);
        assert_eq!(park.values(), (0, 0, 0));

        let drive = ThreeDials::from_gear(Gear::D);
        assert_eq!(drive.values(), (70, 60, 40));

        let sport = ThreeDials::from_gear(Gear::S);
        assert_eq!(sport.values(), (90, 80, 70));
    }

    #[test]
    fn test_three_dials_behavior() {
        let drive = ThreeDials::from_gear(super::super::gear::Gear::D);

        assert!(drive.should_observe());
        assert!(drive.is_observing_actively());
        assert!(drive.should_decide_autonomously());
        assert!(!drive.should_act_immediately()); // 40% < 50%
    }

    #[test]
    fn test_three_dials_parked() {
        let park = ThreeDials::from_gear(super::super::gear::Gear::P);
        assert!(park.is_parked());

        let drive = ThreeDials::from_gear(super::super::gear::Gear::D);
        assert!(!drive.is_parked());
    }

    #[test]
    fn test_autonomy_score() {
        let drive = ThreeDials::from_gear(super::super::gear::Gear::D);
        // (70 + 60 + 40) / 3 = 56.67 -> 56
        assert_eq!(drive.autonomy_score(), 56);
    }

    #[test]
    fn test_matches_gear() {
        use super::super::gear::Gear;

        let drive = ThreeDials::from_gear(Gear::D);
        assert_eq!(drive.matches_gear(), Some(Gear::D));

        let custom = ThreeDials::new(50, 50, 50);
        assert_eq!(custom.matches_gear(), None);
    }
}
